opm/simulators/wells/BlackoilWellModel_impl.hpp

```diff
@@ -1248,42 +1232,73 @@ namespace Opm {
+                                       this->terminal_output_, comm);
 
         // update guide rates
-        const int reportStepIdx = simulator_.episodeIndex();
         if (alq_updated || BlackoilWellModelGuideRates(*this).
-                              guideRateUpdateIsNeeded(reportStepIdx)) {
+                              guideRateUpdateIsNeeded(episodeIdx)) {
             const double simulationTime = simulator_.time();
-            const auto& comm = simulator_.vanguard().grid().comm();
             const auto& summaryState = simulator_.vanguard().summaryState();
             std::vector<Scalar> pot(this->numPhases(), 0.0);
-            const Group& fieldGroup = this->schedule().getGroup("FIELD", reportStepIdx);
+            const Group& fieldGroup = this->schedule().getGroup("FIELD", episodeIdx);
             WellGroupHelpers<Scalar>::updateGuideRates(fieldGroup,
                                                        this->schedule(),
                                                        summaryState,
                                                        this->phase_usage_,
-                                                       reportStepIdx,
+                                                       episodeIdx,
                                                        simulationTime,
                                                        this->wellState(),
                                                        this->groupState(),
                                                        comm,
                                                        &this->guideRate_,
                                                        pot,
-                                                       local_deferredLogger);
+                                                       deferred_logger);
         }
+
         // we need to re-iterate the network when the well group controls changed or gaslift/alq is changed or
         // the inner iterations are did not converge
-        const int iterationIdx = simulator_.model().newtonMethod().numIterations();
-        const bool more_network_update = this->shouldBalanceNetwork(reportStepIdx, iterationIdx) &&
-                    (more_inner_network_update || well_group_control_changed || alq_updated);
-        return {more_network_update, well_group_control_changed};
+        // but after certain number of the iterations, we terminate
+        const std::size_t max_iteration = param_.network_max_outer_iterations_;
+        if (network_update_iteration + 1 == max_iteration) {
+            // only output to terminal if we are at the last newton iterations where we try to balance the network.
+            // the network is not yet balanced
+            const std::string network_str = more_inner_network_update? "": "Network did not converge.";
+            const std::string gaslift_str = alq_updated? "": "Gaslift optimization oscillates.";
```

Comment:
For those strings, You might want a space at the beginning, so the causes are not output without a space between them.